### ISSUE TITLE
Role manager (servo?) assignment behaviour is different when Makecode is/isn't connected

### ISSUE THREAD

**Issue by ross-inksmith** [microsoft/jacdac#1335]
In the attached video (https://github.com/microsoft/jacdac/assets/88154489/bf2d7dc8-3e97-45f6-be6d-22576faf6d69), I've already downloaded the code (https://makecode.microbit.org/_JKyP30b9TY8a) to the microbit. With the computer disconnected I turn the [breakout board](https://microsoft.github.io/jacdac-docs/devices/forward-education/breakoutboardrelaypumpv10/) on. Only the left servo turns (in firmware they're created in order 0:left 1:middle 2:right). Once I connect the computer/Makecode, the left and middle servos start turning. They will keep doing this even if I disconnect the computer again. If I start with the microbit connected to Makecode, then both servos will work on startup.

## Additional detail
In the above build, using our wrapper extension I've included code to wait for `roleManagerServer.isConnected()` in the extension's main.ts before querying (in order) `left-`, `middle-`, and `rightServo.getAngle()` in an attempt to bias the roleManager's assignments. This seems to work in Makecode, but a disconnected microbit will assign "leftServo" to the first service, while the other two roles can be assigned in either order with no obvious pattern.

Our issue has a lot in common with #1293; we can control the servos using the dashboard controls and they move within their full range without delay, but only the first service will respond to the first block (e.g. `servo1.run(x)` will always work).
Other services on the board only start repsonding after the second block (e.g. `servo2.run(x)` has no effect the first time, but subseuqent calls to `servo2.anything` work as normal)

There may be something in the physical init process that influences this; when the board is powered on any servo on the left pin (first one in firmware) moves to 0° while the others do not move. I can turn the servos easily so they're unpowered. Our board doesn't do any detection of servos, so sends signals based on pin regardless of what is plugged in.

Another demonstration is here (https://makecode.microbit.org/_e4UJXP23ER3j) using the MS jacdac extension. If I use the Add Simulators button, it will create three simulators and assign them like this:
![servo-assignment-sims](https://github.com/microsoft/jacdac/assets/88154489/b4df17a2-0876-4b0d-8aa9-609cd2d3f0f6)
Note that the simulators have a 0-180° range but alternate between servo1:0/45° and servo3:45/113° instead of servo1:90/135° and servo3:45/90°
Using the same code and plugging in our breakout board (which has 270° servos), it will now lay itself like this:
![servo-assignment-fwd](https://github.com/microsoft/jacdac/assets/88154489/99a6fafb-4c30-4123-93eb-aad0371bd07e)
The same code now causes the servos to alternate between 0/90° and 90/180° instead of 67.5/135° and 135/202.5°

The above might be related to #1327? I've also found during testing that the starting ranges for our servos initially default to 0-180° before getting updated later.

**Comment by tballmsft** [comment]
Trying to understand. I made a simple example two jacdac buttons - the assignments stay the same when connected and disconnect: https://makecode.microbit.org/_hf6CaPE1Jf3r

**Comment by pelikhan** [comment]
try 3 buttons


Sent from Outlook<http://aka.ms/weboutlook>

________________________________

**Comment by tballmsft** [comment]
There seems to be another issue addressed above:

> Another demonstration is here (https://makecode.microbit.org/_e4UJXP23ER3j) using the MS jacdac extension. If I use the Add Simulators button, it will create three simulators and assign them like this:
> The same code now causes the servos to alternate between 0/90° and 90/180° instead of 67.5/135° and 135/202.5°

The above might be related to https://github.com/microsoft/jacdac/issues/1327? I've also found during testing that the starting ranges for our servos initially default to 0-180° before getting updated later.

**Comment by tballmsft** [comment]
Three buttons seems to work as well: https://makecode.microbit.org/_DYD4RA6rpCK0

**Comment by ross-inksmith** [comment]
The assignment issue does seem to be unique to the servos. I haven't seen it so far with our reflected light/line follower board which also has 3 otherwise identical services on them.

Attached is a packet trace demonstrating the issue similar to #1293, where only one servo responds at first; only one packet is actually sent.

[trace](https://github.com/microsoft/jacdac/files/13031139/trace.jd.1.txt)

**Comment by tballmsft** [comment]
Yes, I am looking now into program you gave above: https://makecode.microbit.org/_e4UJXP23ER3j

**Comment by tballmsft** [comment]
@ross-inksmith, you said:

> There may be something in the physical init process that influences this; when the board is powered on any servo on the left pin (first one in firmware) moves to 0° while the others do not move. I can turn the servos easily so they're unpowered. Our board doesn't do any detection of servos, so sends signals based on pin regardless of what is plugged in.

Right. I forgot that the servos are connected via jacdac cables which PWM is being sent over.  Where is the code that manages initialization?  This must be special-purpose for your board, correct?

**Comment by pelikhan** [comment]
so the difference is that you have 3 servos on the same devices; rather than 3 individual devices with 1 servo service.

**Comment by ross-inksmith** [comment]
> @ross-inksmith, you said:
> 
> > There may be something in the physical init process that influences this; when the board is powered on any servo on the left pin (first one in firmware) moves to 0° while the others do not move. I can turn the servos easily so they're unpowered. Our board doesn't do any detection of servos, so sends signals based on pin regardless of what is plugged in.
> 
> Right. I forgot that the servos are connected via jacdac cables which PWM is being sent over. Where is the code that manages initialization? This must be special-purpose for your board, correct?

Not quite. Similar to the kittenbot device in #1293, we have one jacdac connection to a device with multiple servo services on it. Instead of having it as a separate board, it's on the same board as the edge connector/microbit.
If you were to swap the connector to a standard 3 pin connector, it would work identically

**Comment by ross-inksmith** [comment]
> [trace](https://github.com/microsoft/jacdac/files/13031139/trace.jd.1.txt)

**Comment by pelikhan** [comment]
looks like makecode wants to have 0-180 at start

![image](https://github.com/microsoft/jacdac/assets/4175913/4f5d6769-39b1-4457-a97e-15440979c202)

**Comment by tballmsft** [comment]
@ross-inksmith, you said:

> Not quite. Similar to the kittenbot device in https://github.com/microsoft/jacdac/issues/1293, we have one jacdac connection to a device with multiple servo services on it.

I understand. Where is the firmware for that device?

**Comment by tballmsft** [comment]
@pelikhan - so you think that the makecode defaults for the servo could be leading to the problem?

**Comment by ross-inksmith** [comment]
> @ross-inksmith, you said:
> 
> > Not quite. Similar to the kittenbot device in https://github.com/microsoft/jacdac/issues/1293, we have one jacdac connection to a device with multiple servo services on it.
> 
> I understand. Where is the firmware for that device?

https://github.com/Forward-Education/firmware/tree/main/BreakoutBoard270

**Comment by tballmsft** [comment]
Ok. in jacdac-c servo.c, init() is

```
SRV_DEF(servo, JD_SERVICE_CLASS_SERVO);
void servo_init(const servo_params_t *params) {
    SRV_ALLOC(servo);
    state->params = *params;
    state->params0 = params;
}
```

So nothing happens on init to set the servo to an initial position/state?

**Comment by tballmsft** [comment]
@ross-inksmith  - I see that servo_0 has a different power pin than the other two servos. Correct?

```
const servo_params_t servo_0 = {
    .pin = PA_6,
    .fixed = 0,
    .min_angle = 0 << 16,
    .min_pulse = 600,
    .max_angle = 270 << 16,
    .max_pulse = 2500,
    .power_pin = PB_7
};

const servo_params_t servo_1 = {
    .pin = PA_7,
    .fixed = 0,
    .min_angle = 0 << 16,
    .min_pulse = 600,
    .max_angle = 270 << 16,
    .max_pulse = 2500,
    .power_pin = PC_15
};

const servo_params_t servo_2 = {
    .pin = PA_4,
    .fixed = 0,
    .min_angle = 0 << 16,
    .min_pulse = 600,
    .max_angle = 270 << 16,
    .max_pulse = 2500,
    .power_pin = PC_15
    //NO_PIN
};  
```

**Comment by ross-inksmith** [comment]
@brian-inksmith is double checking the schematics, but the pins are fed from the same rail. (I asked a similar question)

**Comment by tballmsft** [comment]
OK. So buttons generate events, so let me try with something that generates output instead.

**Comment by tballmsft** [comment]
OK. I think I have a repro with the dotMatrix module:

https://makecode.microbit.org/_a2vhjoLqxbEW

![image](https://github.com/microsoft/jacdac/assets/10767146/88e00487-6730-4dea-be9f-96da118db04c)

**Comment by ross-inksmith** [comment]
https://github.com/microsoft/pxt-jacdac/blob/master/servo-server/server.microbit.ts#L13
For the min/max issues, it's also defined here for the servos. It doesn't explain how the math in `servo.run()` comes up with 0/90 and 90/180 though

**Comment by tballmsft** [comment]
Oops spoke too soon... it's working when plugged in.

**Comment by pelikhan** [comment]
The issue is that the servo clients has partially received the ranges from the servo services; and then applies the computation for converting the speed into an angle. Since the current default in the client is -90, 90, you could end up with a combination of [-90,0] x [90, 180].

https://github.com/microsoft/pxt-jacdac/blob/d8afba90caba7f5f0c5c0306a3ee1849083de814/servo/client.ts#L119

**Comment by pelikhan** [comment]
and in restrospect, .... `0 || 90 -> 90` so that explains it.

**Comment by pelikhan** [comment]
https://github.com/microsoft/pxt-jacdac/pull/101

**Comment by ramseyball** [comment]
So the problem is initialization between service and client!

**Comment by ross-inksmith** [comment]
> and in restrospect, .... `0 || 90 -> 90` so that explains it.

Oh that's funny. For the majority of development we were using -90/90
![PXL_20231018_191956597](https://github.com/microsoft/jacdac/assets/88154489/c87aa276-85cd-4901-b006-14dd550237e7)

**Comment by pelikhan** [comment]
silly silly me.

**Comment by pelikhan** [comment]
Servo range fix deployed. Please test.

**Comment by ramseyball** [comment]
Thanks, Peli!

**Comment by ross-inksmith** [comment]
> Servo range fix deployed. Please test.

![neutral](https://github.com/microsoft/jacdac/assets/88154489/276e9f4d-cd51-4563-9352-739d8ee8f481)
![swing](https://github.com/microsoft/jacdac/assets/88154489/f9b16764-bbf0-4fed-82ec-593a4a97eb41)

Also works if I detach the board and add simulators instead :+1:

**Comment by pelikhan** [comment]
I don't think this solves the role manager issue. @ross-inksmith ?

**Comment by ross-inksmith** [comment]
While the angle now reports as accurate, still having issues getting multiple servos to respond. Attached is the video of the exact same code above running without makecode connected. Only servo1 responds. On the first click it goes to it's end range. On the second and subsequent clicks it moves in its normal range. servo3 does not move

https://github.com/microsoft/jacdac/assets/88154489/1475a350-2507-4c9b-9930-77cb273c3a00

**Comment by ross-inksmith** [comment]
Longer video now, behaviour with makecode attached is different. tl;dw: With makecode attached, servo3 will eventually start responding

https://github.com/microsoft/jacdac/assets/88154489/8f018c26-109f-41d7-8bb2-9d4c4d4a3b3a

**Comment by tballmsft** [comment]
OK. This also seems specific to servo.

**Comment by tballmsft** [comment]
In the past, I have seen this sort of behavior where the hardware works differently when the Jacdac device is connected through WebUSB to the dashboard....

**Comment by tballmsft** [comment]
@ross-inksmith you said:

> When I opened the dashboard to record a second trace it worked as with makecode connected but only recorded 18 packets 🤔

Did the program work correctly or not when connected to the dashboard? I'm not sure how to interpret the above comment.

**Comment by tballmsft** [comment]
In the last trace you gave, all the servo messages are tagged with "(usb)":

```
9222.700   b75b1803dc445b97ac759af1133e00e800000000000000000391fc1200736572766f31ff   [ack:0x5bb7] to SB67/PIPE: port:464 cnt:0; sz=19; CMD 0xe800 {     _: 00000000000000000391fc1200736572766f31 | .............servo1 } (usb)
```

**Comment by ross-inksmith** [comment]
Yeah I realized that would be a bit hard to follow so I did it again
[here's the trace.jd.txt](https://github.com/microsoft/jacdac/files/13033682/trace.jd.txt)

and here's a full video
https://github.com/microsoft/jacdac/assets/88154489/5beceef8-bd41-44c3-be64-8a5aa91a2cf6

**Comment by tballmsft** [comment]
@pelikhan - Is it possible to see the log messages that pxt-jacdac sends to the console? That might help more than the Jacdac trace.

**Comment by pelikhan** [comment]
You can turn on the "passive" mode in the jacdac-docs so that the browser does not emit any packets; and you should have the same behavior as without being connected.

![image](https://github.com/microsoft/jacdac/assets/4175913/2e98380b-6ee2-4d04-855f-6c8c7fdbebbc)

**Comment by tballmsft** [comment]
Ah, very good

**Comment by tballmsft** [comment]
I'm not sure this is a role manager problem - seems like the three servos get recognized
```
device: 
  id: TQ01 (0x54cf1386c74a3cdd)
  product: TQ01 (0x3412171b)
  firmware_version: v0.20.3-16-g7e9e119-20230726-1434
  uptime: 72652.17000000224
  stats: announce 103, drop 0, restart 0
  services:
    servo (0x12fc9103)
        const instance_name: "Left" (4c656674) 
            last data: 33726, last get: 33726, last set: , last gets attempts: 0,
        rw angle: 135° (00008700) 
            last data: 80616, last get: 80616, last set: , last gets attempts: 0,
        rw enabled: true (01) 
            last data: 84317, last get: 84317, last set: , last gets attempts: 0,
        rw offset: 0° (00000000) 
            last data: 84322, last get: 84322, last set: , last gets attempts: 0,
        const min_angle: 0° (00000000) 
            last data: 40191, last get: 40191, last set: , last gets attempts: 0,
        const max_angle: 270° (00000e01) 
            last data: 40195, last get: 40195, last set: , last gets attempts: 0,
    servo (0x12fc9103)
        const instance_name: "Middle" (4d6964646c65) 
            last data: 33729, last get: 33729, last set: , last gets attempts: 0,
        rw angle: 0° (00000000) 
            last data: 84571, last get: 84571, last set: , last gets attempts: 0,
        rw enabled: false (00) 
            last data: 84365, last get: 84365, last set: , last gets attempts: 0,
        rw offset: 0° (00000000) 
            last data: 84370, last get: 84370, last set: , last gets attempts: 0,
        const min_angle: 0° (00000000) 
            last data: 34061, last get: 34061, last set: , last gets attempts: 0,
        const max_angle: 270° (00000e01) 
            last data: 34067, last get: 34067, last set: , last gets attempts: 0,
    servo (0x12fc9103)
        const instance_name: "Right" (5269676874) 
            last data: 33740, last get: 33740, last set: , last gets attempts: 0,
        rw angle: 135° (00008700) 
            last data: 80622, last get: 80622, last set: , last gets attempts: 0,
        rw enabled: true (01) 
            last data: 82413, last get: 82413, last set: , last gets attempts: 0,
        rw offset: 0° (00000000) 
            last data: 84414, last get: 84414, last set: , last gets attempts: 0,
        const min_angle: 0° (00000000) 
            last data: 40204, last get: 40204, last set: , last gets attempts: 0,
        const max_angle: 270° (00000e01) 
            last data: 40212, last get: 40212, last set: , last gets attempts: 0,
    relay (0x183fe656)
        const instance_name: "Pump" (50756d70) 
            last data: 33751, last get: 33751, last set: , last gets attempts: 0,
        rw active: false (00) 
            last data: 84264, last get: 84264, last set: , last gets attempts: 0,
```

**Comment by tballmsft** [comment]
My current hypothesis is that we are seeing packet loss. I see something similar here: https://github.com/microsoft/jacdac/issues/1336

**Comment by tballmsft** [comment]
@ross-inksmith - what chip is driving the servos? Is that an STM?

**Comment by brian-forwardedu** [comment]
@tballmsft Yes, specifically it's STM32G030F6P6TR

**Comment by tballmsft** [comment]
Right, just figured that out:

> MCU = STM32g030x6
> include $(PLATFORM)/mk/stm32g0.mk

**Comment by tballmsft** [comment]
@ross-inksmith  - please try swapping the order of the updates to servo1 and servo3 in the program, if you see that servo3 is responding and servo1 is not, then we likely have some packet loss on the STM side.

**Comment by tballmsft** [comment]
You can also try putting a pause between the updates to the two servos.

**Comment by tballmsft** [comment]
Also try this and look in console (when connected over WebUSB)

```
let _ = modules.servo1.angle()
_ = modules.servo2.angle()
_ = modules.servo3.angle()

basic.forever(function () {
    console.log("servo 1 " + modules.servo1.isConnected())
    console.log("servo 2 " + modules.servo2.isConnected())
    console.log("servo 3 " + modules.servo3.isConnected())
    modules.servo1.run(50)
    modules.servo3.run(-50)
    basic.pause(1000)
    modules.servo1.run(0)
    modules.servo3.run(0)
    basic.pause(1000)
})
```

**Comment by tballmsft** [comment]
@ross-inksmith  - I got the servos and am having issues with power on the round board. The green light is on but none of the lights on servos are on. I'm powering the board via its USB port (not Lipo) and power the microbit separately via battery. Any ideas?

**Comment by tballmsft** [comment]
Running the following program (observations follow) on the micro:bit, plugged into the round board, with no servos plugged in:

```
input.onButtonPressed(Button.A, function () {
    modules.servo1.run(50)
    modules.servo3.run(-50)
    basic.pause(1000)
    modules.servo1.run(0)
    modules.servo3.run(0)
    basic.pause(1000)
})

modules.servo1.start()
modules.servo2.start()
modules.servo3.start()

input.onButtonPressed(Button.B, function () {
    console.log("servo 1 " + modules.servo1.isConnected())
    console.log("servo 2 " + modules.servo2.isConnected())
    console.log("servo 3 " + modules.servo3.isConnected())
})

```

When there is no power to the STM, I get
```
binding: 3 servers
servo 1 false
servo 2 false
servo 3 false
```
which is expected.

When I give power to the STM I get
```
binding: 3 servers
servo 1 true
servo 2 true
servo 3 true
```
This is expected, even without servos plugged in, because the STM is hard coded to advertised three servo services, per design, regardless of what is plugged in. So, it seems role management is working fine.

**Comment by tballmsft** [comment]
OK. All three servos are working with the following program (note the display between commands)

```
input.onButtonPressed(Button.A, function () {
    modules.servo1.run(50)
    basic.pause(100)
    modules.servo2.run(50)
    basic.pause(100)
    modules.servo3.run(50)
})

input.onButtonPressed(Button.B, function () {
    modules.servo1.run(0)
    basic.pause(100)
    modules.servo2.run(0)
    basic.pause(100)
    modules.servo3.run(0)
})


modules.servo1.start()
modules.servo2.start()
modules.servo3.start()

input.onButtonPressed(Button.AB, function () {
    console.log("servo 1 " + modules.servo1.isConnected())
    console.log("servo 2 " + modules.servo2.isConnected())
    console.log("servo 3 " + modules.servo3.isConnected())
})
```
So, as I expected, we are losing packets somewhere. Based on the trace files provided before, which showed all expected packets on the bus, my bet is that the STM is dropping packets. But we should analyze more. In the meantime, we can figure out how much of a delay to introduce between commands to different servos.

**Comment by tballmsft** [comment]
OK, now for bringing the delay between run commands down:

-     basic.pause(10)  WORKS!
-

**Comment by tballmsft** [comment]
Note that when micro:bit is connected via USB, first servo often doesn't start.  Seems like more packet loss. Works fine when micro:bit on batter.

**Comment by tballmsft** [comment]
Another thing to consider: if we want to send same command to all three servos, we can use a broadcast, which will use just a single packet.

**Comment by ross-inksmith** [comment]
> @ross-inksmith - I got the servos and am having issues with power on the round board. The green light is on but none of the lights on servos are on. I'm powering the board via its USB port (not Lipo) and power the microbit separately via battery. Any ideas?
> 
> Wait a second, the green light is on, with the switch in the OFF position. When I switch it ON, the led turns red.
> 
> I am very confused. I think we need a call.

There are two LEDs by the port on the board. One is green to indicate a charge, the other is a low battery/low voltage light. When driving multiple power hungry devices (two servos and a LED ring, for example), or if the servos draw extra current to create more torque, or if some combination of those things happen when the battery isn't fully charged, that can also turn the red light on.

And that's right, the board only works when switched on regardless of whether or not it's charging - it only draws power from the battery or port; it won't charge through the microbit/edge connector.

> So, as I expected, we are losing packets somewhere. Based on the trace files provided before, which showed all expected packets on the bus, my bet is that the STM is dropping packets. But we should analyze more. In the meantime, we can figure out how much of a delay to introduce between commands to different servos.

We were wondering if this was an issue in #1312, which is also STM

**Comment by pelikhan** [comment]
Try again; fix pushed.

**Comment by tballmsft** [comment]
It is working now - Peli made pxt-jacdac/servo last chatty - please update to latest pxt-jacdac

**Comment by tballmsft** [comment]
This was not a problem with the STM - pxt-jacdac was too chatty.

**Comment by brian-inksmith** [comment]
Initial tests seem good.  Continuing to test.

**Comment by ross-inksmith** [comment]
I'm unexpectedly out of office today and can't test with our devices today. Reading the diff looks cool though

Creating a new project and adding the jacdac extension includes v1.9.19, but using our extension loads a cached version (in this case, v1.9.13). Emptying the cache via dev tools doesn't get it either. I can only get it by through chrome://settings/content/all?search=data 

Unfortunately, a lot of classrooms use chromebooks that don't have access to chrome settings or incognito mode. Is there another way? I don't see anything on https://makecode.com/extensions/pxt-json or in https://github.com/microsoft/pxt/blob/master/pxtlib/package.ts

**Comment by pelikhan** [comment]
how do you publish release of your extension? do you use `mkc bump`? The mkc cli bumps the cloud to flush its caches otherwise the makecode cloud has a 24h cache for extensions.

**Comment by ross-inksmith** [comment]
I do use `mkc bump`. I don't think the cloud cache is the issue, since it would use the current version of jacdac in incognito and the issue went away by clearing local data. I've created an issue with pxt https://github.com/microsoft/pxt/issues/9740

**Comment by pelikhan** [comment]
@ross-inksmith can you share again, it won't load

**Comment by ross-inksmith** [comment]
sorry, which?

**Comment by ross-inksmith** [comment]
I think it broke because I cleaned up the branch it was using. I've made this to the best of my memory https://makecode.microbit.org/_df5Udy6drWiE. It uses the version of our extension from two weeks ago but jacdac v1.9.20

**Comment by pelikhan** [comment]
In 1.9.23, the `?srvo=yy` hint is now used when binding roles. So you'll want to add those to your role names as such:

https://github.com/microsoft/pxt-jacdac/blob/master/servo/test.microbit.ts#L4

This gives me a stable assignment with 1, 2 or 3 servos when the shield is connected. Let's how far we go with this. Eventually we could also add the product id hint; it's just that i might not have at hand when assigning roles.

**Comment by ross-inksmith** [comment]
This is excellent! Working as advertised for me so far

**Comment by pelikhan** [comment]
Let's see how far you get with testing with that fix.

**Comment by ramseyball** [comment]
If there are other problems, please file a separate issue for each problem.
